src/pudl/extract/eiaaeo.py

```diff
@@ -0,0 +1,252 @@
+
+    date: Timestamp = Field(coerce=True)
+    value: str = Field(coerce=True)
+    units: str = Field(coerce=True)
+    series_name: str = Field(coerce=True)
+    category_name: str = Field(coerce=True)
+    case: str = Field(coerce=True)
+
+
+class AEOTaxonomy:
+    """Container for *all* the information in one AEO report."""
+
+    def __init__(self, records: Iterable[str]):
+        """Load AEO JSON records into a graph datastructure.
+
+        Args:
+            records: the strings that contain an AEO report to parse.
+
+        """
+        categories, series = self.__load_records(records)
+        self.graph = self.__generate_graph(categories, series)
+        root = 4949903
+        self.__cases = set(self.graph.successors(root))
+        self.__sanitize_re = re.compile(r"\W+")
+
+    def __load_records(
+        self, records: Iterable[str]
+    ) -> tuple[dict[int, Category], dict[str, Series]]:
+        # A record can be either a category or a series, so we parse those into
+        # two separate mappings.
+
+        all_categories: dict[int, Category] = {}
+        all_series: dict[str, Series] = {}
+        for record in records:
+            if "category_id" in record:
+                category = Category.model_validate_json(record)
+                all_categories[category.category_id] = category
+            elif "series_id" in record:
+                series = Series.model_validate_json(record)
+                all_series[series.series_id] = series
+            else:
+                raise ValueError(f"Line had neither series nor category ID: {record}")
+        return all_categories, all_series
+
+    def __generate_graph(
+        self, categories: dict[int, Category], series: dict[str, Series]
+    ) -> nx.DiGraph:
+        def get_all_edges(category: Category) -> Iterable[tuple[int, str | int]]:
+            # A category can have incoming edges from its parents or outgoing
+            # edges to its child series.
+            edges: list[tuple[int, str | int]] = [
+                (category.parent_category_id, category.category_id)
+            ]
+            if category.childseries:
+                edges += [
+                    (category.category_id, child_series_id)
+                    for child_series_id in category.childseries
+                ]
+            return edges
+
+        edges = itertools.chain.from_iterable(
+            get_all_edges(category) for category in categories.values()
+        )
+        graph = nx.DiGraph(incoming_graph_data=edges)
+
+        nx.set_node_attributes(graph, categories | series)
+        return graph
+
+    def __sanitize(self, s: str) -> str:
+        return re.sub(self.__sanitize_re, "_", s.lower().strip().replace(" : ", "__"))
+
+    def __series_to_records(
+        self, series_id: str, potential_parents: set[int]
+    ) -> pd.DataFrame:
+        # Use the graph info to figure out what case + category name this
+        # series belongs to
```

Comment:
I realize these are all private methods, but is there a reason not to use docstrings rather than comments to document them? Then the documentation will pop up automatically in editors & notebooks for our own reference.